text/0000-ssr-rehydration.md

```diff
@@ -0,0 +1,299 @@
+import { someSideEffect } from './my-app';
+
+export default class Example extends LightningServerElement {
+    @api message = 'Hello World!';
+
+    onClick () {
+      someSideEffect();
+    }
+}
+```
+
+**Server Build:**
+
+```javascript
+import { api } from 'lwc';
+import { LightningServerElement } from 'some-framework';
+import { someSideEffect } from './my-app';
+
+export default class Example extends LightningServerElement {
+    @api message = 'Hello World!';
+
+    // LightningServerElement hooks into component lifecycle:
+    //   connectedCallback () {
+    //    this.setAttribute('data-ssr-preserve', true);
+    //   }
+
+    onClick () {
+      someSideEffect();
+    }
+}
+```
+
+**SSR HTML:**
+
+```html
+<!-- ... -->
+  <x-example data-ssr-preserve>
+    <template shadowroot="open">
+      Hello World!
+    </template>
+  </x-example>
+<!-- ... -->
+```
+
+**Client Build:**
+
+```javascript
+import { api } from 'lwc';
+import { NoopElement } from 'some-framework';
+import { someSideEffect } from './my-app'; // <-- dead-code, removed with minification
+
+// The entire class declaration is replaced.
+export default class Example extends NoopElement {}
+```
+
+</details>
+
+### SSR Placeholders
+
+Implementation of placeholder components, as described in [Exception A to the happy path](#the-happy-path) should be practicable without changes to LWC, so long as unexpected subtree mismatches (Exception C) are dealt with properly.
+
+Implementation in user-land is possible with simple runtime environment detection:
+
+<details open>
+  <summary>Example:</summary>
+
+```javascript
+import { LightningElement } from 'lwc';
+import { isNodeEnv } from './my-utils';
+import tmplServerPlaceholder from './templateServer.html';
+import tmplClient from './templateClient.html';
+
+export default class ElementWithSSRPlaceholder extends LightningElement {
+    render() {
+        return this.isNodeEnv ? tmplServerPlaceholder : tmplClient;
+    }
+}
+```
+</details><br />
+
+Alternately, if it is desirable _not_ to ship the placeholder in the client-side bundle, one could utilize a Babel transform or `@rollup/plugin-replace` to make the necessary changes at build time.
+
+<details open>
+  <summary>Example:</summary>
+
+```javascript
+import { LightningElement } from 'lwc';
+import { isNodeEnv } from './my-utils';
+// ↶ removed from client builds during minification
+import tmplServerPlaceholder from './templateServer.html';
+import tmplClient from './templateClient.html';
+
+export default class ElementWithSSRPlaceholder extends LightningElement {
+    render() {
+        /*
+          On the server, the following line is transformed to:
+            return 'server' === 'server'
+          On the client, it is transformed to:
+            return 'client' === 'server'
+        */
+        return process.env.BUILD_TARGET === 'server'
+            ? tmplServerPlaceholder
+            : tmplClient;

```

Comment:
The convention I've seen for this in the past is [`process.browser`](https://www.npmjs.com/package/process#browser-implementation). There is also [the `"browser"` field in `package.json`](https://github.com/defunctzombie/package-browser-field-spec) and more recently [conditional exports](https://nodejs.org/api/packages.html#packages_conditional_exports).

It seems to me we should align on `process.browser` since it's no weirder than `process.env.NODE_ENV === 'production'` (which we already emit in our build). Also, anything that relies on `package.json` results in trickier code for authors to write (need two separate files for browser vs Node code), and for us to support in Rollup/Webpack/Jest/etc. If another convention has gained more traction or seems more appropriate, though (`import.meta.browser`?) then I'm happy to align on that instead.

Another alternative is to force users to rely on sniffing (e.g. [`browser-or-node`](https://www.npmjs.com/package/browser-or-node)), but the downside is that we can't strip it out in Rollup/minification.